### ISSUE TITLE
Split-KV decode rounds partial outputs to the output dtype, in a workspace already sized for fp32

### ISSUE THREAD

**Issue by Wint3rNight** [flashinfer-ai/flashinfer#5107]
### Before submitting

- [x] I have searched existing issues and did not find a solution or an existing report.

### Environment

```text
python -m flashinfer.collect_env
```

### Bug description

When the decode scheduler splits the KV context into chunks, each chunk's partial output goes into `tmp_v` and is merged afterwards. `tmp_v` is allocated at `sizeof(float)` per element but read back as `DTypeO*`, so every partial is rounded to bf16 before the merge. The upper half of the allocation is never touched, so the precision is given up without buying any memory back.

Where it comes from:

- `include/flashinfer/attention/scheduler.cuh:485` allocates `num_qo_heads * padded_batch_size * HEAD_DIM * sizeof(float)`
- `csrc/batch_decode.cu:231` reads that offset back as `DTypeO*`
- `include/flashinfer/attention/decode.cuh:610` does `st.o.cast_store(...)`, so the fp32 accumulator is rounded on the way out

For the config in the repro below that is 512 KiB allocated and 256 KiB actually written.

Prefill has the same shape: `scheduler.cuh:860` also allocates with `sizeof(float)` (`sizeof_dtype_o` is passed in but discarded at line 775), and `csrc/batch_prefill.cu:210` and `:346` read it back as `DTypeO*`.

The effect is that with split-KV active, a third to a half of decode outputs are off by at least one bf16 step against an fp32 reference, and turning the split off with everything else held constant drops that to 0.2 percent.

### Minimal reproduction

```python
import torch, flashinfer
torch.manual_seed(0)
dev = "cuda:0"
HQ, HKV, D, KV_LEN, PAGE = 32, 4, 128, 16384, 16
DT = torch.bfloat16
npages = KV_LEN // PAGE

q1 = torch.randn(1, HQ, D, device=dev, dtype=DT)
kv = (torch.randn(npages, 2, PAGE, HKV, D, device=dev, dtype=torch.float32) * 0.5).to(DT)

# fp32 reference for one request, from the same bf16 inputs
k = kv[:, 0].reshape(-1, HKV, D).float()
v = kv[:, 1].reshape(-1, HKV, D).float()
g = HQ // HKV
ref = torch.empty(HQ, D, device=dev, dtype=torch.float32)
for h in range(HQ):
    s = (q1[0, h].float() @ k[:, h // g].T) / (D ** 0.5)
    ref[h] = torch.softmax(s, dim=-1) @ v[:, h // g]
ref_bf16 = ref.to(DT)

for B in (1, 2, 4, 8, 16, 32, 64, 128):
    # all requests share the same pages, so the math is identical and memory stays flat
    kv_indptr = torch.arange(0, B + 1, device=dev, dtype=torch.int32) * npages
    kv_indices = torch.arange(npages, device=dev, dtype=torch.int32).repeat(B)
    last = torch.full((B,), PAGE, device=dev, dtype=torch.int32)
    ws = torch.empty(512 * 1024 * 1024, dtype=torch.int8, device=dev)
    w = flashinfer.decode.BatchDecodeWithPagedKVCacheWrapper(ws, "NHD", use_tensor_cores=False)
    w.plan(kv_indptr, kv_indices, last, HQ, HKV, D, PAGE, data_type=DT, q_data_type=DT)
    o = w.run(q1.expand(B, HQ, D).contiguous(), kv)
    pi = w._plan_info
    bad = (o[0] != ref_bf16).sum().item()
    print(f"batch {B:>3}  split_kv={pi[9]}  padded={pi[0]:>3}  "
          f"mismatches {bad:>4}/{ref_bf16.numel()} ({100.0 * bad / ref_bf16.numel():5.1f}%)")



batch   1  split_kv=1  padded= 32  mismatches 2097/4096 ( 51.2%)
batch   2  split_kv=1  padded= 32  mismatches 1999/4096 ( 48.8%)
batch   4  split_kv=1  padded= 32  mismatches 1659/4096 ( 40.5%)
batch   8  split_kv=1  padded= 32  mismatches 1664/4096 ( 40.6%)
batch  16  split_kv=1  padded= 32  mismatches 1502/4096 ( 36.7%)
batch  32  split_kv=0  padded= 32  mismatches    8/4096 (  0.2%)
batch  64  split_kv=0  padded= 64  mismatches    8/4096 (  0.2%)
batch 128  split_kv=0  padded=128  mismatches    8/4096 (  0.2%)


A mismatch means the returned bf16 value is not the correctly rounded bf16 of the fp32 reference, so each one is off by at least one bf16 step. Same query, same KV cache, same kernel across every row. The only thing that changes is whether the scheduler decided to split, and that decision flips between batch 16 and batch 32. Small batch with long context and GQA is exactly the regime where the split always fires.

Side note for anyone reproducing this the obvious way: `plan(..., disable_split_kv=True)` will not help, because it is only forwarded on the tensor-core branch (`decode.py:1264`) and never reaches `DecodePlanImpl`, so it is silently ignored when `use_tensor_cores=False`. Varying the batch size is why the loop above does it that way.
```

### Error messages / logs

```text

```

### Additional context

There are two ways to resolve this and they exclude each other.

1. Write the partials as fp32. Recovers the precision with no change to the allocation, since the buffer is already that size. The cost is bandwidth on `tmp_v`: 512 KiB extra on top of a 32 MiB KV read for the config above, about 1.6 percent.
2. #4356, which sizes `tmp_v` by the output dtype. Reclaims the unused half and makes the current rounding the contract.

Both are reasonable, and there is precedent for 2-byte partials elsewhere in the tree (`holistic_partial_o` and `mla_partial_o` are explicitly sized at 2 bytes with an "assume fp16" note). But if 2 is the intent, then split decode is knowingly giving up about a bf16 step on a third to a half of its outputs in the small-batch long-context case, and that seems worth writing down rather than leaving for someone to find.

Which would you rather have? I can put up the PR for either one.

**Comment by XFDG** [comment]
Hi, I've submitted a fix in https://github.com/flashinfer-ai/flashinfer/pull/5166.
